Add require decls to Interface and NamedConstraint

danakj opened the pull request:
They are not used for impl lookup or verifying anything yet, but now they appear in the textual semir.

danakj committed:
require-in-interface

danakj commented:
This is based on https://github.com/carbon-language/carbon-lang/pull/6318, the first new commit is `require-in-interface`

jonmeow commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -1341,6 +1341,18 @@ struct RequireCompleteType {
+          {.ir_name = "require_decl", .is_lowered = false});
+
+  // Always `type`.
+  TypeId type_id;
```
I see there's a little similar with interfaces and named constraints (decls have type type), but my instinct would usually be that the type should be separate (e.g. a singleton, or a class decl vs class type split which may preserve more info for diagnostics). What made you choose this particular typing approach?

jonmeow replied:
Closing in favor of https://github.com/carbon-language/carbon-lang/pull/6318#discussion_r2491852239

jonmeow commented on toolchain/sem_ir/interface.h:
```diff
@@ -24,6 +24,9 @@ struct InterfaceFields {
 
   // The following members are set at the `}` of the interface definition.
+
+  // TODO: Should this be a block of `RequireImplsId` to avoid an indirection?
```
FWIW I think it might be nice if that's enforced, assuming it's always true. I think of the use of instructions as a mixture of convenience and flexibility for inserting other things, rather than a strong preference towards generic storage.

Under such an approach, would the `RequireDecl` instruction still be needed, or could that go onto the `RequireImplsId` fully?

danakj replied:
I think the RequiresDecl is needed to connect the instructions inside the `require` to their enclosing scope in the semir.

What I wasn't sure about with a `RequireImplsId` set here is if it would be a vector in place here, which would be convenient but is not what we do for insts - we have an Id here. So I would need to add another type of Id for RequireImplsBlockId. So let's leave it as a TODO for now, but if you have opinions about this LMK.

jonmeow commented on toolchain/sem_ir/inst_namer.cpp:
```diff
@@ -1129,6 +1143,26 @@ auto InstNamer::NamingContext::NameInst() -> void {
+      }
+      // FIXME: Do this?
+      // AddBlockLabel(scope_id, inst.decl_block_id, "require",
+      //                            SemIR::LocId(inst_id_));
```
Missed cleaning this up? IIRC this is mainly for if we start having things like conditions within blocks, e.g. supporting `if x then y else z` which is currently allowed in statement expressions but not declarations (but probably should work more broadly). So you may not observe it now, but similar to `class` adding a label, my guess is it's probably for the better to not get lost later.

```suggestion
      AddBlockLabel(scope_id, inst.decl_block_id, "require",
                                 SemIR::LocId(inst_id_));
```

danakj replied:
Ah okay, thanks. Will do.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -1159,6 +1177,12 @@ auto Formatter::FormatInstRhs(Inst inst) -> void {
+      FormatArgs(decl.require_impls_id);
+      FormatTrailingBlock(decl.decl_block_id);
+      return;
+    }
```
nit: maybe move this before the `Return` instructions, for lexical sorting?

danakj replied:
oops, yes

jonmeow commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -243,6 +243,16 @@ struct Worklist {
+    if (!require_id.has_value()) {
+      AddInvalid();
+      return;
+    }
```
This seems a little unusual, when is it the case that you'd fingerprint a `RequireDecl` without an id associated?

i.e., wondering why this isn't:

```suggestion
    CARBON_CHECK(require_id.has_value());
```

danakj replied:
It just makes it flexible so it doesn't have to be changed later. There's nothing _bad_ about a None being formatted in the future. But it can be CHECK for now, yeah.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -413,6 +413,16 @@ auto Formatter::FormatInterface(InterfaceId id, const Interface& interface_info)
+      FormatArg(
+          sem_ir_->insts().GetAs<SemIR::RequireDecl>(inst_id).require_impls_id);
+    }
+    out_ << "\n";
```
Here and below, would it make sense to elide these when empty?

(separately, are these long enough to be worth a `FormatRequiresBlock` helper?)

jonmeow replied:
Note, mainly it feels odd to me that the common case is `{{}}`. Perhaps the formatting of `RequireDecl` could change instead to make it more clear what's going on?

danakj replied:
I went back and forth on making these be elided when empty. It's nice maybe for _this_ PR to avoid all the semir churn. But it means adding/removing a `require` can change more lines of semir than otherwise (granted its already with a change there). Ultimately I landed on that I thought it was nice to show in the semir that there is a section for requires even if it's empty for a given interface or constraint. Whereas for other scopes, there is not one.

That said, the `{{}}` is weird, and not what I intended and was not like that when I first wrote this, I put the `Indent()` in the wrong place when I moved things around to not make it conditional. Fixed.

jonmeow commented on toolchain/check/eval_inst.cpp:
```diff
@@ -516,6 +516,17 @@ auto EvalConstantInst(Context& context, SemIR::InstId inst_id,
+      .type_id = inst.type_id,
+      .require_impls_id = inst.require_impls_id,
+      // The constant value doesn't include the body block so that we don't try
+      // to import it, and it doesn't pollute the textual semir.
```
```suggestion
      // to import it, and it doesn't pollute the textual SemIR.
```

nit: TBH I find the capitalization helpful, it means you aren't referring to the "semir" object in specific, but rather the more abstract SemIR. Either way, I try to capitalize proper nouns/acronyms in comments.

danakj replied:
Done

danakj committed:
check

danakj committed:
format

danakj committed:
FormatRequireImplsBlock

danakj reviewed (commented):
PTAL (note I will be renaming RequireDecl to RequireImplsDecl in the dependent PR).

danakj commented:
PTAL

I rebased this on https://github.com/carbon-language/carbon-lang/pull/6318 because some of the changes suggested here were really part of that PR, and have been applied there in the meantime, and things got renamed and moved around a fair bit over there. That said, no changes needed to happen in this PR other than renames in the process.

The first commit is still/again `require-in-interface`.

jonmeow commented on toolchain/check/context.h:
```diff
@@ -354,9 +354,10 @@ class Context {
+  // param_and_arg_refs_stack_. It's used for:
+  // - struct literals, where we need to track names for a type separate from
+  //   the literal arguments.
+  // - the associated entries witness table, while parsing an interface.
```
```suggestion
  // - Struct literals, where we need to track names for a type separate from
  //   the literal arguments.
  // - The associated entries witness table, while parsing an interface.
```

Nit: particularly with the period on both, it seems like you should use sentence capitalization.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -417,6 +417,10 @@ auto Formatter::FormatInterface(InterfaceId id, const Interface& interface_info)
     out_ << "\n";
 
+    IndentLabel();
+    out_ << "!requires:\n";
```
It seems like this could be part of `FormatRequireImplsBlock` too; I get that it's outside on `!members`, but also `FormatNameScope` is used for other blocks.

danakj replied:
I kind of like having the top level things all written here, but sure, done.

jonmeow commented on toolchain/check/handle_interface.cpp:
```diff
@@ -165,12 +165,19 @@ auto HandleParseNode(Context& context, Parse::InterfaceDefinitionId /*node_id*/)
+  llvm::SmallVector<SemIR::InstId> require_decls(llvm::make_filter_range(
+      context.inst_blocks().Get(inst_block_id), [&](SemIR::InstId inst_id) {
+        return context.insts().Is<SemIR::RequireImplsDecl>(inst_id);
+      }));
```
Looking at this again, why are these filtered at the end instead of being collected during construction?

e.g., with something like arguments, we accumulate them as part of the context. Couldn't these require decls be accumulated similarly (and more efficiently)?

danakj replied:
If we add another stack to the Context, we can accumulate them directly into that stack and then pop them here.

It turns out putting these Decl insts into the Interface is painful for importing them, so I think I will switch the storage to `RequireImplsIds`, which means I will have to add that stack. So, done!

danakj committed:
caps

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

danakj committed:
require-blocks

jonmeow reviewed (approved):
Approving, generally LG on the changes in this PR.

To note though, regarding the `RequireImplsDecl` instruction being added in the base PR... Particularly as you say there's import complexity, I'd wonder if you're going to end up wanting to drop it on import. If you do, perhaps drop that instruction entirely?

Mainly I say that because even though having the SemIR show up as it does might be nice, on import we'd need to reconcile the `RequireImplsId` to match. That's probably going to be a bit complex, to make the interface import and instruction import mirror, since the ids don't have a direct meaning.

I'm also looking at removing the pending specific/generic code because of problems arising from it, instead making them run through the work queue. I note that just in case it affects what you're doing. :)

But really I don't think that would affect the delta in this PR (maybe a merge conflict, but now this PR pretty much uses the `RequireImpls` storage without the decl).

danakj commented:
> Approving, generally LG on the changes in this PR.
> 
> To note though, regarding the `RequireImplsDecl` instruction being added in the base PR... Particularly as you say there's import complexity, I'd wonder if you're going to end up wanting to drop it on import. If you do, perhaps drop that instruction entirely?

The import is done in https://github.com/carbon-language/carbon-lang/pull/6344 now, and I think the instruction is needed because:
1. The import needs to be multi-phase. It needs to make placeholder ids and then import dependent things, and then fill in those ids.
2. To be multi-phase it needs to store the result in a ConstantId for an InstId. Making import support being multi-phase over a RequireImplsId would be a very big lift.

danakj commented:
> > Approving, generally LG on the changes in this PR.
> > To note though, regarding the `RequireImplsDecl` instruction being added in the base PR... Particularly as you say there's import complexity, I'd wonder if you're going to end up wanting to drop it on import. If you do, perhaps drop that instruction entirely?
> 
> The import is done in #6344 now, and I think the instruction is needed because:
> 
> 1. The import needs to be multi-phase. It needs to make placeholder ids and then import dependent things, and then fill in those ids.
> 2. To be multi-phase it needs to store the result in a ConstantId for an InstId. Making import support being multi-phase over a RequireImplsId would be a very big lift.

I tried to cut out the decl instruction anyway, as I think we can reorder things to drop a phase. So could we drop the decl inst then? But then we have to create a Generic without a decl, which is also weird.
